cmd/ib-sriov-cni/main.go

```diff
@@ -164,15 +160,82 @@ func cmdAdd(args *skel.CmdArgs) error {
 		}()
 	}
 
-	result := &current.Result{}
-	result.Interfaces = []*current.Interface{{
-		Name:    args.IfName,
-		Sandbox: netns.Path(),
-	}}
-
 	err = sm.SetupVF(netConf, args.IfName, args.ContainerID, netns)
+	if err != nil {
+		return fmt.Errorf("failed to set up pod interface %q from the device %q: %v", args.IfName, netConf.Master, err)
+	}
+	return nil
+}
+
+// Run the IPAM plugin
+func runIPAMplugin(args *skel.CmdArgs, netConf *localtypes.NetConf, netns ns.NetNS, result **current.Result) error {
+	if netConf.IPAM.Type == ipamDHCP {
+		return fmt.Errorf("ipam type dhcp is not supported")
+	}
+	r, err := ipam.ExecAdd(netConf.IPAM.Type, args.StdinData)
+	if err != nil {
+		return fmt.Errorf("failed to set up IPAM plugin type %q from the device %q: %v",
+			netConf.IPAM.Type, netConf.Master, err)
+	}
+
+	defer func() {
+		if err != nil {
+			_ = ipam.ExecDel(netConf.IPAM.Type, args.StdinData)
+		}
+	}()
+
+	// Convert the IPAM result into the current Result type
+	var newResult *current.Result
+	newResult, err = current.NewResultFromResult(r)
+	if err != nil {
+		return err
+	}
+
+	if len(newResult.IPs) == 0 {
+		return errors.New("IPAM plugin returned missing IP config")
+	}
+
+	newResult.Interfaces = (*result).Interfaces
+
+	for _, ipc := range newResult.IPs {
+		// All addresses apply to the container interface (move from host)
+		ipc.Interface = current.Int(0)
+	}
+
+	err = netns.Do(func(_ ns.NetNS) error {
+		return ipam.ConfigureIface(args.IfName, newResult)
+	})
+	if err != nil {
+		return err
+	}
+	*result = newResult
+	return nil
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	netConf, netns, err := getNetConfNetns(args)
+	if err != nil {
+		return err
+	}
+	defer netns.Close()
+
+	sm := sriov.NewSriovManager()
+
+	// Lock CNI operation to serialize the operation
+	lock, err := lockCNIExecution()
+	if err != nil {
+		return err
+	}
+	defer unlockCNIExecution(lock)
+
+	err = doVFConfiguration(sm, netConf, netns, args)
 	defer func() {
 		if err != nil {
+			// restore RDMA device back to default namespace in case of error
+			// Note(adrianc): as there is no logging, we have little visibility if the restore operation failed.
+			if netConf.RdmaIso {
+				_ = utils.MoveRdmaDevFromNs(netConf.RdmaNetState.ContainerRdmaDevName, netns)
```

Comment:
this cleanup will happen twice if `doVFConfiguration` fails, once in the defer in L#160 and here

or will trigger the cleanup here even if we failed early in `doVFConfiguration` before RDMA device was moved to namespace.

Reply:
if the cleanup is assured to do nothing when ran more than once
or when ran when it shouldn't have then im ok with it if it keeps the code cleaner.